### ISSUE TITLE
Kong fails to detect sibling pods accurately during shutdown

### ISSUE THREAD

**Issue by seh** [Kong/kubernetes-ingress-controller#581]
## Summary

When Kong shuts down and [the `--update-status-on-shutdown` command-line flag](https://github.com/Kong/kubernetes-ingress-controller/blob/0.7.1/cli/ingress-controller/flags.go#L176-L178) is set to true, Kong tries to figure out whether it's the last Kong instance running in [the `internal/ingress/status.statusSync.Shutdown` function](https://github.com/Kong/kubernetes-ingress-controller/blob/0.7.1/internal/ingress/status/status.go#L106-L149).

While it does a reasonable job finding sibling pods in some cases, it fails to do so for pods supervised by _ReplicaSets_ and their parent _Deployments_. This failure to find actual sibling pods causes Kong to clear the "status" subresource of _Ingress_ objects. Observers of such objects such as External DNS then conclude that the _Ingress_ is no longer serving, and take action unnecessarily.

**Kong Ingress controller version**

[v0.7.1](https://github.com/Kong/kubernetes-ingress-controller/releases/tag/0.7.1)

**Kong or Kong Enterprise version**

Kong (non-Enterprise)

### Kubernetes version

`version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.12", GitCommit:"a8b52209ee172232b6db7a6e0ce2adc77458829f", GitTreeState:"clean", BuildDate:"2019-10-15T12:04:30Z", GoVersion:"go1.11.13", Compiler:"gc", Platform:"linux/amd64"}`

### Environment

- **Cloud provider or hardware configuration**:
AWS EC2 _m5.xlarge_ instance
- **OS** (e.g. from /etc/os-release):
```
PRETTY_NAME="Debian GNU/Linux 9 (stretch)"
NAME="Debian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```
- **Kernel** (e.g. `uname -a`):
`Linux ip-10-3-105-163 4.9.0-9-amd64 #1 SMP Debian 4.9.168-1+deb9u5 (2019-08-11) x86_64 GNU/Linux`
- **Install tools**:
_kops_

## What happened

Create a Kubernetes _Deployment_ running the Kong ingress controller with its `--update-status-on-shutdown` flag set to true. Scale the _Deployment_ to use one replica. Ensure that one Kong instance is healthy, and is currently the elected leader for updating _Ingress_ "status" subresources.

Change the _Deployment_ in some way (such as changing a pod label) that requires introducing a new _ReplicaSet_, or reactivating an older one. Configure either the _Depyloment_'s update strategy or a governing _PodDisruptionBudget_ to keep the original, leading Kong pod running until the replacement pod is ready. Once that new pod is ready, the kubelet hosting the original pod will send its containers SIGTERM, and Kong will initiate its shutdown procedure.

Now, assuming that Kong is sitting behind a _Service_ of type "LoadBalancer," and doesn't have [its `--publish-status-address` command-line flag](https://github.com/Kong/kubernetes-ingress-controller/blob/0.7.1/cli/ingress-controller/flags.go#L170-L173) set, [`syncStatus.runningAddresses`](https://github.com/Kong/kubernetes-ingress-controller/blob/0.7.1/internal/ingress/status/status.go#L208-L259) will return a single address. With [only one address in hand, Kong moves on to look for sibling pods](https://github.com/Kong/kubernetes-ingress-controller/blob/0.7.1/internal/ingress/status/status.go#L136-L145) with its alternate technique: [the `syncStatus.isRunningMultiplePods` function](https://github.com/Kong/kubernetes-ingress-controller/blob/0.7.1/internal/ingress/status/status.go#L271-L280).

How does `isRunningMultiplePods` find candidate sibling pods? Earlier Kong had captured the _current pod's labels_ discovered via [the `utils.GetPodDetails` function](https://github.com/Kong/kubernetes-ingress-controller/blob/0.7.1/internal/ingress/utils/k8s.go#L84-L105), using the downward API to identify itself. Using those labels, it [searches for pods in the same namespace](https://github.com/Kong/kubernetes-ingress-controller/blob/0.7.1/internal/ingress/status/status.go#L272-L274) **with those same labels**. If there's more than one such pod (omitting itself), there must be siblings.

Recall that these pods were created transitively by a _Deployment_. A _Deployment_ [sets the "pod-template-hash" label](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#pod-template-hash-label) on both the _ReplicaSets_ it creates and in the pod template inside each of these _ReplicaSets_. That "pod-template-hash" label value will differ between the two _ReplicatSets_ that created the first running Kong pod and its replacements (for example, due to a change in the pods' labels). Capturing the Kong pod's labels includes this "pod-template-hash" label. Searching for pods with the same "pod-template-hash" label **omits sibling pods created by different _ReplicaSets_**, even though those other pods will serve behind the same Kubernetes _Service_ inspected earlier. Even if there were no _Service_ involved—such as when serving on host ports—those other pods are still part of the intended set of Kong pods.

## Expected behavior

Kong should either discover these sibling pods by using alternate criteria, or declare that it can't be sure and take no action.

### Inspect a fronting _Service_'s pod selector

If Kong is running behind a Kubernetes _Service_ [with a pod selector](https://kubernetes.io/docs/concepts/services-networking/service/#services-without-selectors), use _that_ selector to find all candidate Kong pods. That selector will usually be more general than the labels on any matching pods, forming the union of pods created by several _ReplicaSets_ at a time.

### Inspect a fronting _Service_'s companion _Endpoints_ object

Since Kubernetes is already maintaining the ready set of pods for a _Service_ with a pod selector in an _Endpoints_ object, we could consult that instead of querying with a pod selector ourselves.

### Accept a label selector as a command-line flag

Rather than mistakenly inferring that the current pod's labels are a fitting selector for its siblings, honor a command-line flag that allows the operator to tell Kong explicitly how to find these pods.

**Comment by hbagdi** [comment]
Hey @seh,
Thanks for the detailed description here. I'm surprised how this bug is lurking around for so long without anybody else running into it. Good find!

Regarding the solutions proposed:

> Inspect a fronting Service's pod selector
> Inspect a fronting Service's companion Endpoints object

These two will not work in every deployment topology. While, the most common deployment has the controller and kong containers in the same pod, there are other deployments where the controller and kong containers are running in separate pods. In these deployments, if we inspect the Service's endpoint, it will lead to an incorrect pods being identified as the controller pods.

We can take an input of label-selector as a CLI flag, but that seems like additional burden to maintain.
The labels such as `pod-template-hash` are not very common. Would it be an acceptable solution to ignore the `pod-template-hash` label and working with the remaining labels in the pod definition?

**Comment by seh** [comment]
I hadn't thought about that separation of containers into multiple pods. I agree that the _Service_ would be pointing elsewhere there.

You can ignore the "pod-template-hash" label and help a few cases, but I still don't think it's the safe way to proceed. Adding the proposed flag puts the behavior in the operator's hands, and that's the only party who knows the right value to use. You could make the default value of the selector be the current pod's labels minus "pod-template-hash", if present, but I still think the documentation should advise setting that flag value to something different.

**Comment by hbagdi** [comment]
> You can ignore the "pod-template-hash" label and help a few cases, but I still don't think it's the safe way to proceed. Adding the proposed flag puts the behavior in the operator's hands, and that's the only party who knows the right value to use. You could make the default value of the selector be the current pod's labels minus "pod-template-hash", if present, but I still think the documentation should advise setting that flag value to something different.

That works!
Are you open to contributing this feature?

**Comment by seh** [comment]
> Are you open to contributing this feature?

Yes, I am. I need to survey the code again to get a feel for how hard it's going to be. I'll probably need pointers for rounding out a change like this that touches documentation.

**Comment by seh** [comment]
As I get in deeper here, is there any reason to preserve [the current inspection of the _Service_ fronting Kong](https://sourcegraph.com/github.com/Kong/kubernetes-ingress-controller/-/blob/internal/ingress/status/status.go#L217-L234) to determine whether there are several ingress controller containers? Just because [a load balancer has multiple IP addresses available](https://sourcegraph.com/github.com/Kong/kubernetes-ingress-controller/-/blob/internal/ingress/status/status.go#L130-L140) does not imply that there is more than one container hosting a Kong ingress controller. A load balancer may manifest as a single IP address but still point to many targets, or it could present several IP addresses (covering multiple availability zones) that all still lead to a single container in a single pod.

This looks like we're performing extra work while leaving us vulnerable to both false positives and negatives.

**Comment by hbagdi** [comment]
I see the problem here. I think that code is making an assumption of running as a NodePort deployment.
Isn't the original issue here and the one you are referring to two different issues though? If so, let's track them in two different issues.

**Comment by seh** [comment]
Note, though, that [we're explicitly looking for _Service_s of type "LoadBalancer"](https://sourcegraph.com/github.com/Kong/kubernetes-ingress-controller/-/blob/internal/ingress/status/status.go#L224), which subsume the capabilities of those of type "NodePort". [The "LoadBalancer" field in a _Service_'s status](https://github.com/kubernetes/api/blob/2433a9db3db38d5e177eac8495dec1cd3d15b128/core/v1/types.go#L3825-L3828) is only populated for those of type "LoadBalancer".

If this was a summary of a _Service_ of type "NodePort", and Kubernetes exposed a set of pertinent addresses (which it does not), finding multiple addresses would _still_ be misleading, because that's representative of the nodes in the cluster, not the number of selected pods in the cluster.

This point is related to the original complaint here, because this first predicate determines whether the ingress controller carries on to try to list pods. It can falsely assume that there are multiple controllers in play when there are not. The original complaint here focuses on the converse—assuming there are fewer sibling pods than there really are—but the consequence is also incorrect behavior, just in a different direction.

**Comment by seh** [comment]
Working through this on Friday, I noticed that the `status.runningAddresses` function also assumes that the ingress controller and proxy run together in the same pod. That's sometimes true, but not always true.

I ran out of time for the day to pursue it further, but it occurred to me that we probably need not one but two label selectors here: one to find proxy pods, and one to find ingress controller pods. By default we can make the former assume the latter's value, but a single label selector won't do for cases where the proxy and ingress controller are separated.

**Comment by hbagdi** [comment]
Steven,

Thanks for the detailed comments.
I see the mess here. Status part has received absolutely no love from our side. This is all reminiscent code from the ingress-nginx codebase when Alejandro ported it over to Kong.
I agree with mostly everything you're saying.

> Note, though, that we're explicitly looking for Services of type "LoadBalancer", which subsume the capabilities of those of type "NodePort".

That is only the first case. If you look at the default case, it takes into account the NodePort type service.

> but it occurred to me that we probably need not one but two label selectors here: one to find proxy pods, and one to find ingress controller pods.

We already take in `--publish-service` and can easily find out the associated endpoints for the proxy endpoint using this info. In case the pods are running outside k8s, then `--publish-status-address` can help. Does this help?

> This point is related to the original complaint here, because this first predicate determines whether the ingress controller carries on to try to list pods. It can falsely assume that there are multiple controllers in play when there are not. The original complaint here focuses on the converse—assuming there are fewer sibling pods than there really are—but the consequence is also incorrect behavior, just in a different direction.

Right. We have to solve both the problems. Can we solve them one at a time is what I'm asking for? If you feel you can tackle both of them in the same PR, that works for me as well.
My recommendation is here to first proceed with solving the original issue in this thread using the explicit input of labels for detecting controller pods.

**Comment by mflendrich** [comment]
This is being reworked in KIC 2.0.
